Flickr/Flickr/ImageCollectionViewController.swift

```diff
@@ -105,114 +147,255 @@ class ImageCollectionViewController: UIViewController,  UICollectionViewDelegate
+            startEditingState()
+    }
+    
+    @objc func refreshCurrentCollectionViewByPullToUpdate() {
+        print("reload")
+        pageFlickr = 1
+        isNotUpdating = true
+        if self.collectionViewSearch.alpha == 0 {
+            self.performFlickrPopular()
+            return
+        } else {
+            guard let lastValue = self.lastEnteredTextValue else {
+                return
+            }
+            self.performFlickrSearch(url: lastValue)
+        }
     }
     
     @objc func clickOnTextEventFunc(textField: UITextField) {
-        searchHistoryView.reloadData()
-        searchHistoryShowMustGoOn()
+        updateHistoryList()
+        startEditingState()
     }
     
     func setDelegates_DataSources() {
-        searchTextField.delegate = self
-        searchHistoryView.delegate = self
-        searchHistoryView.dataSource = self
+        textSearch.delegate = self
+        tableViewHistorySearch.delegate = self
+        tableViewHistorySearch.dataSource = self
+    }
+    
+    func setMagnifyAndCancelButtonAlphasDefault() {
+        buttonCancel.alpha = 0
+        imageMagnify.alpha = 0.5
     }
     
     func setAlphasDefault() {
         subViewForSpinner.alpha = 0
-        searchCollectionView.alpha = 0
-        popularCollectionView.alpha = 1
-        searchHistoryView.alpha = 0
+        collectionViewSearch.alpha = 0
+        collectionViewPopular.alpha = 1
+        tableViewHistorySearch.alpha = 0
+        setMagnifyAndCancelButtonAlphasDefault()
+    }
+    
+    func scrollBack(view: UICollectionView, indexPath: IndexPath) {
+        view.scrollToItem(at: indexPath, at: .centeredVertically, animated: false)
     }
     
     func setConstraintMode() {
-        searchConstraint.priority = UILayoutPriority(rawValue: 999)
-        searchConstraint.isActive = true
+        tableViewHistorySearch.clipsToBounds = true
+        constraintForHideTextField.priority = UILayoutPriority(rawValue: 999)
+        heightConstraintTableView.constant = 0
+        heightConstraintTableView.isActive = true
+        constraintForHideTextField.isActive = true
+    }
+    
+    @objc func tapImageAction(recognizer: UITapGestureRecognizer) {
+        textSearch.becomeFirstResponder()
+        startEditingState()
+    }
+    
+    func hideAllCollectionViews() {
+        self.collectionViewSearch.alpha = 0
+        self.collectionViewPopular.alpha = 0
+    }
+    
+    func showEditingStyle() {
+        self.imageMagnify.alpha = 1
+        self.buttonCancel.alpha = 1
+        self.tableViewHistorySearch.alpha = 1
+    }
+    
+    func startEditingState() {
+        UIView.animate(withDuration: ConstantNumbers.standartTimeForAnimation,  animations: {
+            self.hideAllCollectionViews()
+            self.showEditingStyle()
+        })
+        rebuildTableSize()
+    }
+    
+    func isCollectionsViewHidedBoth() -> Bool {
+        let check = (collectionViewSearch.alpha == 0) && (collectionViewPopular.alpha == 0)
```

Comment:
getting logic info from UI is wrong. Separate global bool variable should be created